diff --git a/NEWS b/NEWS
index a7950c73b..87b4cf93f 100644
--- a/NEWS
+++ b/NEWS
@@ -1,3 +1,9 @@
+OVN v24.03.6 - xx xxx xxxx
+--------------------------
+   - Improved handling of IPv6 traffic by enabling address prefix tracking
+     in OVS for both IPv4 and IPv6 addresses, whenever possible, reducing
+     the amount of IPv6 datapath flows.
+
 OVN v24.03.5 - 21 Jan 2025
 --------------------------
   - Bug fixes
diff --git a/TODO.rst b/TODO.rst
index 17d539ad6..01370e38b 100644
--- a/TODO.rst
+++ b/TODO.rst
@@ -130,3 +130,6 @@ OVN To-do List
   * Move the lflow build parallel processing from northd.c to lflow-mgr.c
     This would also ensure that the variables declared in northd.c
     (eg. thread_lflow_counter) are not externed in lflow-mgr.c.
+
+* Remove OVS database schema parsing from features.c once server side column
+  types are available through IDL.
diff --git a/configure.ac b/configure.ac
index 5a36a0057..96e7d5ca6 100644
--- a/configure.ac
+++ b/configure.ac
@@ -13,7 +13,7 @@
 # limitations under the License.
 
 AC_PREREQ(2.63)
-AC_INIT(ovn, 24.03.5, bugs@openvswitch.org)
+AC_INIT(ovn, 24.03.6, bugs@openvswitch.org)
 AC_CONFIG_MACRO_DIR([m4])
 AC_CONFIG_AUX_DIR([build-aux])
 AC_CONFIG_HEADERS([config.h])
diff --git a/controller/mac_cache.c b/controller/mac_cache.c
index 43b68fae0..cb66c3827 100644
--- a/controller/mac_cache.c
+++ b/controller/mac_cache.c
@@ -36,6 +36,13 @@ mac_cache_mac_binding_find(struct mac_cache_data *data,
 static bool
 mac_cache_mb_data_from_sbrec(struct mac_cache_mb_data *data,
                               const struct sbrec_mac_binding *mb);
+static void mac_cache_mac_binding_update_log(
+    const char *action,
+    const struct mac_cache_mb_data *,
+    bool print_times,
+    const struct mac_cache_threshold *,
+    int64_t idle_age_ms,
+    uint64_t since_updated_ms);
 static uint32_t
 mac_cache_fdb_data_hash(const struct mac_cache_fdb_data *fdb_data);
 static inline bool
@@ -78,7 +85,16 @@ mac_cache_threshold_add(struct mac_cache_data *data,
     threshold = xmalloc(sizeof *threshold);
     threshold->uuid = dp->header_.uuid;
     threshold->value = value;
-    threshold->dump_period = (3 * value) / 4;
+    threshold->dump_period = value / 2;
+    threshold->cooldown_period = value / 4;
+
+    /* (cooldown_period + dump_period) is the maximum time the timestamp may
+     * be not updated.  So, the sum of those times must be lower than the
+     * threshold, otherwise we may fail to update an active MAC binding in
+     * time and risk it being removed.  Giving it an extra 1/10 of the time
+     * for all the processing that needs to happen. */
+    ovs_assert(threshold->cooldown_period + threshold->dump_period
+               < (9 * value) / 10);
 
     hmap_insert(thresholds, &threshold->hmap_node,
                 uuid_hash(&dp->header_.uuid));
@@ -130,6 +146,8 @@ mac_cache_mac_binding_add(struct mac_cache_data *data,
     }
 
     mc_mb->data = mb_data;
+    mc_mb->sbrec = mb;
+    mac_cache_mac_binding_update_log("Added", &mb_data, false, NULL, 0, 0);
 }
 
 void
@@ -147,26 +165,12 @@ mac_cache_mac_binding_remove(struct mac_cache_data *data,
         return;
     }
 
+    mac_cache_mac_binding_update_log("Removed", &mc_mb->data, false,
+                                     NULL, 0, 0);
     hmap_remove(&data->mac_bindings, &mc_mb->hmap_node);
     free(mc_mb);
 }
 
-bool
-mac_cache_sb_mac_binding_updated(const struct sbrec_mac_binding *mb)
-{
-    bool updated = false;
-    for (size_t i = 0; i < SBREC_MAC_BINDING_N_COLUMNS; i++) {
-        /* Ignore timestamp update as this does not affect the existing nodes
-         * at all. */
-        if (i == SBREC_MAC_BINDING_COL_TIMESTAMP) {
-            continue;
-        }
-        updated |= sbrec_mac_binding_is_updated(mb, i);
-    }
-
-    return updated || sbrec_mac_binding_is_deleted(mb);
-}
-
 void
 mac_cache_mac_bindings_clear(struct mac_cache_data *data)
 {
@@ -176,27 +180,31 @@ mac_cache_mac_bindings_clear(struct mac_cache_data *data)
     }
 }
 
+
+static void
+mac_cache_mac_binding_data_to_string(const struct mac_cache_mb_data *data,
+                                     struct ds *out_data)
+{
+    char ip[INET6_ADDRSTRLEN];
+
+    if (!ipv6_string_mapped(ip, &data->ip)) {
+        return;
+    }
+    ds_put_format(out_data, "cookie: 0x%08"PRIx64", "
+                            "datapath-key: %"PRIu32", "
+                            "port-key: %"PRIu32", "
+                            "ip: %s, mac: " ETH_ADDR_FMT,
+                  data->cookie, data->dp_key, data->port_key,
+                  ip, ETH_ADDR_ARGS(data->mac));
+}
+
 void
 mac_cache_mac_bindings_to_string(const struct hmap *map, struct ds *out_data)
 {
     struct mac_cache_mac_binding *mb;
     HMAP_FOR_EACH (mb, hmap_node, map) {
-        char ip[INET6_ADDRSTRLEN];
-
-        if (!ipv6_string_mapped(ip, &mb->data.ip)) {
-            continue;
-        }
-        if (mb->data.sbrec) {
-            ds_put_format(out_data, "SB UUID: " UUID_FMT ", ",
-                          UUID_ARGS(&mb->data.sbrec->header_.uuid));
-        } else {
-            ds_put_cstr(out_data, "SB UUID: <none>, ");
-        }
-        ds_put_format(out_data, "datapath-key: %"PRIu32", "
-                                "port-key: %"PRIu32", "
-                                "ip: %s, mac: " ETH_ADDR_FMT "\n",
-                      mb->data.dp_key, mb->data.port_key,
-                      ip, ETH_ADDR_ARGS(mb->data.mac));
+        mac_cache_mac_binding_data_to_string(&mb->data, out_data);
+        ds_put_char(out_data, '\n');
     }
 }
 
@@ -239,22 +247,6 @@ mac_cache_fdb_remove(struct mac_cache_data *data, const struct sbrec_fdb *fdb)
     free(mc_fdb);
 }
 
-bool
-mac_cache_sb_fdb_updated(const struct sbrec_fdb *fdb)
-{
-    bool updated = false;
-    for (size_t i = 0; i < SBREC_FDB_N_COLUMNS; i++) {
-        /* Ignore timestamp update as this does not affect the existing nodes
-         * at all. */
-        if (i == SBREC_FDB_COL_TIMESTAMP) {
-            continue;
-        }
-        updated |= sbrec_fdb_is_updated(fdb, i);
-    }
-
-    return updated || sbrec_fdb_is_deleted(fdb);
-}
-
 void
 mac_cache_fdbs_clear(struct mac_cache_data *data)
 {
@@ -285,7 +277,9 @@ mac_cache_mb_stats_process_flow_stats(struct ovs_list *stats_list,
 
     stats->idle_age_ms = ofp_stats->idle_age * 1000;
     stats->data.mb = (struct mac_cache_mb_data) {
-        .port_key = ofp_stats->match.flow.regs[MFF_LOG_INPORT - MFF_REG0],
+        .cookie = ntohll(ofp_stats->cookie),
+        /* The port_key must be zero to match mac_binding_data_from_sbrec. */
+        .port_key = 0,
         .dp_key = ntohll(ofp_stats->match.flow.metadata),
         .mac = ofp_stats->match.flow.dl_src
     };
@@ -299,6 +293,34 @@ mac_cache_mb_stats_process_flow_stats(struct ovs_list *stats_list,
     ovs_list_push_back(stats_list, &stats->list_node);
 }
 
+static void
+mac_cache_mac_binding_update_log(
+    const char *action,
+    const struct mac_cache_mb_data *data,
+    bool print_times,
+    const struct mac_cache_threshold *threshold,
+    int64_t idle_age_ms, uint64_t since_updated_ms)
+{
+    if (!VLOG_IS_DBG_ENABLED()) {
+        return;
+    }
+
+    struct ds s = DS_EMPTY_INITIALIZER;
+
+    ds_put_cstr(&s, action);
+    ds_put_cstr(&s, " MAC binding (");
+    mac_cache_mac_binding_data_to_string(data, &s);
+    if (print_times) {
+        ds_put_format(&s, "), last update: %"PRIu64"ms ago,"
+                          " idle age: %"PRIi64"ms, threshold: %"PRIu64"ms",
+                      since_updated_ms, idle_age_ms, threshold->value);
+    } else {
+        ds_put_char(&s, ')');
+    }
+    VLOG_DBG("%s.", ds_cstr_ro(&s));
+    ds_destroy(&s);
+}
+
 void
 mac_cache_mb_stats_run(struct ovs_list *stats_list, uint64_t *req_delay,
                        void *data)
@@ -312,27 +334,48 @@ mac_cache_mb_stats_run(struct ovs_list *stats_list, uint64_t *req_delay,
         struct mac_cache_mac_binding *mc_mb =
                 mac_cache_mac_binding_find(cache_data, &stats->data.mb);
         if (!mc_mb) {
+            mac_cache_mac_binding_update_log("Not found in the cache:",
+                                             &stats->data.mb, false,
+                                             NULL, 0, 0);
             free(stats);
             continue;
         }
 
-        struct uuid *dp_uuid = &mc_mb->data.sbrec->datapath->header_.uuid;
+        struct uuid *dp_uuid = &mc_mb->sbrec->datapath->header_.uuid;
+        uint64_t since_updated_ms = timewall_now - mc_mb->sbrec->timestamp;
         struct mac_cache_threshold *threshold =
                 mac_cache_threshold_find(thresholds, dp_uuid);
 
         /* If "idle_age" is under threshold it means that the mac binding is
-         * used on this chassis. Also make sure that we don't update the
-         * timestamp more than once during the dump period. */
-        if (stats->idle_age_ms < threshold->value &&
-            (timewall_now - mc_mb->data.sbrec->timestamp) >=
-            threshold->dump_period) {
-            sbrec_mac_binding_set_timestamp(mc_mb->data.sbrec, timewall_now);
+         * used on this chassis. */
+        if (stats->idle_age_ms < threshold->value) {
+            if (since_updated_ms >= threshold->cooldown_period) {
+                mac_cache_mac_binding_update_log("Updating active",
+                                                 &mc_mb->data, true, threshold,
+                                                 stats->idle_age_ms,
+                                                 since_updated_ms);
+                sbrec_mac_binding_set_timestamp(mc_mb->sbrec, timewall_now);
+            } else {
+                /* Postponing the update to avoid sending database transactions
+                 * too frequently. */
+                mac_cache_mac_binding_update_log("Not updating active",
+                                                 &mc_mb->data, true, threshold,
+                                                 stats->idle_age_ms,
+                                                 since_updated_ms);
+            }
+        } else {
+            mac_cache_mac_binding_update_log("Not updating non-active",
+                                             &mc_mb->data, true, threshold,
+                                             stats->idle_age_ms,
+                                             since_updated_ms);
         }
-
         free(stats);
     }
 
     mac_cache_update_req_delay(thresholds, req_delay);
+    if (*req_delay) {
+        VLOG_DBG("MAC binding statistics dalay: %"PRIu64, *req_delay);
+    }
 }
 
 void
@@ -351,6 +394,34 @@ mac_cache_fdb_stats_process_flow_stats(struct ovs_list *stats_list,
     ovs_list_push_back(stats_list, &stats->list_node);
 }
 
+static void
+mac_cache_fdb_update_log(const char *action,
+                         const struct mac_cache_fdb_data *data,
+                         bool print_times,
+                         const struct mac_cache_threshold *threshold,
+                         int64_t idle_age_ms, uint64_t since_updated_ms)
+{
+    if (!VLOG_IS_DBG_ENABLED()) {
+        return;
+    }
+
+    struct ds s = DS_EMPTY_INITIALIZER;
+
+    ds_put_cstr(&s, action);
+    ds_put_format(&s, " FDB entry (datapath-key: %"PRIu32", "
+                      "port-key: %"PRIu32", mac: " ETH_ADDR_FMT,
+                  data->dp_key, data->port_key, ETH_ADDR_ARGS(data->mac));
+    if (print_times) {
+        ds_put_format(&s, "), last update: %"PRIu64"ms ago,"
+                          " idle age: %"PRIi64"ms, threshold: %"PRIu64"ms",
+                      since_updated_ms, idle_age_ms, threshold->value);
+    } else {
+        ds_put_char(&s, ')');
+    }
+    VLOG_DBG("%s.", ds_cstr_ro(&s));
+    ds_destroy(&s);
+}
+
 void
 mac_cache_fdb_stats_run(struct ovs_list *stats_list, uint64_t *req_delay,
                         void *data)
@@ -364,25 +435,46 @@ mac_cache_fdb_stats_run(struct ovs_list *stats_list, uint64_t *req_delay,
         struct mac_cache_fdb *mc_fdb = mac_cache_fdb_find(cache_data,
                                                           &stats->data.fdb);
         if (!mc_fdb) {
+            mac_cache_fdb_update_log("Not found in the cache:",
+                                     &stats->data.fdb, false, NULL, 0, 0);
             free(stats);
             continue;
         }
 
+        uint64_t since_updated_ms =
+            timewall_now - mc_fdb->sbrec_fdb->timestamp;
         struct mac_cache_threshold *threshold =
                 mac_cache_threshold_find(thresholds, &mc_fdb->dp_uuid);
-        /* If "idle_age" is under threshold it means that the mac binding is
-         * used on this chassis. Also make sure that we don't update the
-         * timestamp more than once during the dump period. */
-        if (stats->idle_age_ms < threshold->value &&
-            (timewall_now - mc_fdb->sbrec_fdb->timestamp) >=
-            threshold->dump_period) {
-            sbrec_fdb_set_timestamp(mc_fdb->sbrec_fdb, timewall_now);
+        /* If "idle_age" is under threshold it means that the fdb entry is
+         * used on this chassis. */
+        if (stats->idle_age_ms < threshold->value) {
+            if (since_updated_ms >= threshold->cooldown_period) {
+                mac_cache_fdb_update_log("Updating active",
+                                         &mc_fdb->data, true, threshold,
+                                        stats->idle_age_ms,
+                                        since_updated_ms);
+                sbrec_fdb_set_timestamp(mc_fdb->sbrec_fdb, timewall_now);
+            } else {
+                /* Postponing the update to avoid sending database transactions
+                 * too frequently. */
+                mac_cache_fdb_update_log("Not updating active",
+                                         &mc_fdb->data, true, threshold,
+                                         stats->idle_age_ms,
+                                         since_updated_ms);
+            }
+        } else {
+            mac_cache_fdb_update_log("Not updating non-active",
+                                     &mc_fdb->data, true, threshold,
+                                     stats->idle_age_ms, since_updated_ms);
         }
 
         free(stats);
     }
 
     mac_cache_update_req_delay(thresholds, req_delay);
+    if (*req_delay) {
+        VLOG_DBG("FDB entry statistics dalay: %"PRIu64, *req_delay);
+    }
 }
 
 void
@@ -397,9 +489,7 @@ mac_cache_stats_destroy(struct ovs_list *stats_list)
 static uint32_t
 mac_cache_mb_data_hash(const struct mac_cache_mb_data *mb_data)
 {
-    uint32_t hash = mb_data->sbrec
-                    ? uuid_hash(&mb_data->sbrec->header_.uuid)
-                    : 0;
+    uint32_t hash = hash_uint64(mb_data->cookie);
 
     hash = hash_add(hash, mb_data->port_key);
     hash = hash_add(hash, mb_data->dp_key);
@@ -413,7 +503,7 @@ static inline bool
 mac_cache_mb_data_equals(const struct mac_cache_mb_data *a,
                           const struct mac_cache_mb_data *b)
 {
-    return a->sbrec == b->sbrec &&
+    return a->cookie == b->cookie &&
            a->port_key == b->port_key &&
            a->dp_key == b->dp_key &&
            ipv6_addr_equals(&a->ip, &b->ip) &&
@@ -451,7 +541,7 @@ mac_cache_mb_data_from_sbrec(struct mac_cache_mb_data *data,
         return false;
     }
 
-    data->sbrec = mb;
+    data->cookie = mb->header_.uuid.parts[0];
     return true;
 }
 
diff --git a/controller/mac_cache.h b/controller/mac_cache.h
index 8705b63e6..ee5096670 100644
--- a/controller/mac_cache.h
+++ b/controller/mac_cache.h
@@ -47,10 +47,12 @@ struct mac_cache_threshold {
     uint64_t value;
     /* Statistics dump period. */
     uint64_t dump_period;
+    /* How long to wait between two database updates. */
+    uint64_t cooldown_period;
 };
 
 struct mac_cache_mb_data {
-    const struct sbrec_mac_binding *sbrec; /* Only the UUID is used as key.*/
+    uint64_t cookie;
     uint32_t port_key;
     uint32_t dp_key;
     struct in6_addr ip;
@@ -61,6 +63,8 @@ struct mac_cache_mac_binding {
     struct hmap_node hmap_node;
     /* Common data to identify MAC binding. */
     struct mac_cache_mb_data data;
+    /* Reference to the SB MAC binding record (Might be NULL). */
+    const struct sbrec_mac_binding *sbrec;
 };
 
 struct mac_cache_fdb_data {
@@ -93,7 +97,7 @@ mac_cache_mac_binding_data_init(struct mac_cache_mb_data *data,
                                 struct in6_addr ip, struct eth_addr mac)
 {
     *data = (struct mac_cache_mb_data) {
-        .sbrec = NULL,
+        .cookie = 0,
         .dp_key = (dp_key),
         .port_key = (port_key),
         .ip = (ip),
@@ -112,13 +116,10 @@ void mac_cache_mac_bindings_clear(struct mac_cache_data *data);
 void mac_cache_mac_bindings_to_string(const struct hmap *map,
                                       struct ds *out_data);
 
-bool mac_cache_sb_mac_binding_updated(const struct sbrec_mac_binding *mb);
-
 void mac_cache_fdb_add(struct mac_cache_data *data,
                        const struct sbrec_fdb *fdb, struct uuid dp_uuid);
 void mac_cache_fdb_remove(struct mac_cache_data *data,
                           const struct sbrec_fdb *fdb);
-bool mac_cache_sb_fdb_updated(const struct sbrec_fdb *fdb);
 void mac_cache_fdbs_clear(struct mac_cache_data *data);
 
 void
diff --git a/controller/ovn-controller.c b/controller/ovn-controller.c
index 10b4c5573..a03aed801 100644
--- a/controller/ovn-controller.c
+++ b/controller/ovn-controller.c
@@ -487,6 +487,76 @@ create_br_datapath(struct ovsdb_idl_txn *ovs_idl_txn,
     return dp;
 }
 
+#define N_FLOW_TABLES 255
+
+static void
+update_flow_table_prefixes(struct ovsdb_idl_txn *ovs_idl_txn,
+                           const struct ovsrec_bridge *br_int)
+{
+    size_t max_prefixes = ovs_features_max_flow_table_prefixes_get();
+    struct ds ds = DS_EMPTY_INITIALIZER;
+    const char *prefixes[] = {
+        "ip_src", "ip_dst", "ipv6_src", "ipv6_dst",
+    };
+    struct ovsrec_flow_table *ft;
+    size_t i;
+
+    /* We must not attempt setting more prefixes than our IDL supports.
+     * Note: This should be a build time assertion, but IDL structures
+     * are not defined as constants. */
+    ovs_assert(
+        ARRAY_SIZE(prefixes) <=
+        ovsrec_flow_table_columns[OVSREC_FLOW_TABLE_COL_PREFIXES].type.n_max);
+
+    if (!max_prefixes) {
+        /* Not discovered yet. */
+        return;
+    }
+
+    max_prefixes = MIN(max_prefixes, ARRAY_SIZE(prefixes));
+    if (br_int->n_flow_tables == N_FLOW_TABLES &&
+        br_int->value_flow_tables[0]->n_prefixes == max_prefixes) {
+        /* Already up to date.  Ideally, we would check every table,
+         * but it seems excessive. */
+        return;
+    }
+
+    for (i = 1; i < br_int->n_flow_tables; i++) {
+        if (br_int->value_flow_tables[i] != br_int->value_flow_tables[0]) {
+            break;
+        }
+    }
+    if (i == N_FLOW_TABLES) {
+        /* Correct number of flow tables and all pointing to the same row. */
+        ft = br_int->value_flow_tables[0];
+    } else {
+        /* Unexpected configuration.  Let's create a new flow table row.
+         * Old ones will be garbage collected by the database. */
+        struct ovsrec_flow_table *values[N_FLOW_TABLES];
+        int64_t keys[N_FLOW_TABLES];
+
+        ft = ovsrec_flow_table_insert(ovs_idl_txn);
+        for (i = 0; i < ARRAY_SIZE(values); i++) {
+            keys[i] = i;
+            values[i] = ft;
+        }
+        ovsrec_bridge_set_flow_tables(br_int, keys, values,
+                                      ARRAY_SIZE(values));
+    }
+
+    ds_put_cstr(&ds, "Setting flow table prefixes:");
+    for (i = 0 ; i < max_prefixes; i++) {
+        ds_put_char(&ds, ' ');
+        ds_put_cstr(&ds, prefixes[i]);
+        ds_put_char(&ds, ',');
+    }
+    ds_chomp(&ds, ',');
+    VLOG_INFO("%s.", ds_cstr_ro(&ds));
+    ds_destroy(&ds);
+
+    ovsrec_flow_table_set_prefixes(ft, prefixes, max_prefixes);
+}
+
 static const struct ovsrec_bridge *
 get_br_int(const struct ovsrec_bridge_table *bridge_table,
            const struct ovsrec_open_vswitch_table *ovs_table)
@@ -563,6 +633,8 @@ process_br_int(struct ovsdb_idl_txn *ovs_idl_txn,
                                                     datapath_type);
                 }
             }
+
+            update_flow_table_prefixes(ovs_idl_txn, br_int);
         }
     }
     *br_int_ = br_int;
@@ -1112,8 +1184,11 @@ ctrl_register_ovs_idl(struct ovsdb_idl *ovs_idl)
     ovsdb_idl_add_column(ovs_idl, &ovsrec_bridge_col_ports);
     ovsdb_idl_add_column(ovs_idl, &ovsrec_bridge_col_name);
     ovsdb_idl_add_column(ovs_idl, &ovsrec_bridge_col_fail_mode);
+    ovsdb_idl_add_column(ovs_idl, &ovsrec_bridge_col_flow_tables);
     ovsdb_idl_add_column(ovs_idl, &ovsrec_bridge_col_other_config);
     ovsdb_idl_add_column(ovs_idl, &ovsrec_bridge_col_external_ids);
+    ovsdb_idl_add_table(ovs_idl, &ovsrec_table_flow_table);
+    ovsdb_idl_add_column(ovs_idl, &ovsrec_flow_table_col_prefixes);
     ovsdb_idl_add_table(ovs_idl, &ovsrec_table_ssl);
     ovsdb_idl_add_column(ovs_idl, &ovsrec_ssl_col_bootstrap_ca_cert);
     ovsdb_idl_add_column(ovs_idl, &ovsrec_ssl_col_ca_cert);
@@ -3399,10 +3474,6 @@ mac_cache_sb_mac_binding_handler(struct engine_node *node, void *data)
 
     const struct sbrec_mac_binding *sbrec_mb;
     SBREC_MAC_BINDING_TABLE_FOR_EACH_TRACKED (sbrec_mb, mb_table) {
-        if (!mac_cache_sb_mac_binding_updated(sbrec_mb)) {
-            continue;
-        }
-
         if (!sbrec_mac_binding_is_new(sbrec_mb)) {
             mac_cache_mac_binding_remove(cache_data, sbrec_mb);
         }
@@ -3440,10 +3511,6 @@ mac_cache_sb_fdb_handler(struct engine_node *node, void *data)
     struct local_datapath *local_dp;
     const struct sbrec_fdb *sbrec_fdb;
     SBREC_FDB_TABLE_FOR_EACH_TRACKED (sbrec_fdb, fdb_table) {
-        if (!mac_cache_sb_fdb_updated(sbrec_fdb)) {
-            continue;
-        }
-
         if (!sbrec_fdb_is_new(sbrec_fdb)) {
             mac_cache_fdb_remove(cache_data, sbrec_fdb);
         }
@@ -5185,6 +5252,12 @@ main(int argc, char *argv[])
                          &sbrec_chassis_private_col_nb_cfg);
     ovsdb_idl_omit_alert(ovnsb_idl_loop.idl,
                          &sbrec_chassis_private_col_nb_cfg_timestamp);
+    /* Omit the timestamp columns of the MAC_Binding and FDB tables.
+     * ovn-controller doesn't need to react to changes in timestamp
+     * values (it does read them to implement aging).  Therefore we
+     * can disable change tracking and alerting for these columns. */
+    ovsdb_idl_omit_alert(ovnsb_idl_loop.idl, &sbrec_mac_binding_col_timestamp);
+    ovsdb_idl_omit_alert(ovnsb_idl_loop.idl, &sbrec_fdb_col_timestamp);
 
     /* Omit the external_ids column of all the tables except for -
      *  - DNS. pinctrl.c uses the external_ids column of DNS,
@@ -5726,7 +5799,8 @@ main(int argc, char *argv[])
             if (ovs_idl_txn
                 && ovs_feature_support_run(br_int_dp ?
                                            &br_int_dp->capabilities : NULL,
-                                           br_int ? br_int->name : NULL)) {
+                                           br_int ? br_int->name : NULL,
+                                           ovs_remote)) {
                 VLOG_INFO("OVS feature set changed, force recompute.");
                 engine_set_force_recompute(true);
 
diff --git a/controller/pinctrl.c b/controller/pinctrl.c
index e8069f106..c1ba1cfb9 100644
--- a/controller/pinctrl.c
+++ b/controller/pinctrl.c
@@ -4320,6 +4320,9 @@ mac_binding_add_to_sb(struct ovsdb_idl_txn *ovnsb_idl_txn,
         /* For backward compatibility check if timestamp column is available
          * in SB DB. */
         if (pinctrl.mac_binding_can_timestamp) {
+            VLOG_DBG("Setting MAC binding timestamp for "
+                     "ip:%s mac:%s port:%s to %lld",
+                     b->ip, b->mac, logical_port, time_wall_msec());
             sbrec_mac_binding_set_timestamp(b, time_wall_msec());
         }
     }
diff --git a/debian/changelog b/debian/changelog
index 9b569314e..280a806d3 100644
--- a/debian/changelog
+++ b/debian/changelog
@@ -1,3 +1,9 @@
+ovn (24.03.6-1) unstable; urgency=low
+   [ OVN team ]
+   * New upstream version
+
+ -- OVN team <dev@openvswitch.org>  Tue, 21 Jan 2025 18:05:28 -0500
+
 ovn (24.03.5-1) unstable; urgency=low
    [ OVN team ]
    * New upstream version
diff --git a/ic/ovn-ic.c b/ic/ovn-ic.c
index 12ad69489..ccea9424e 100644
--- a/ic/ovn-ic.c
+++ b/ic/ovn-ic.c
@@ -1123,6 +1123,8 @@ add_to_routes_ad(struct hmap *routes_ad, const struct in6_addr prefix,
                  const struct nbrec_logical_router_static_route *nb_route,
                  const struct nbrec_logical_router *nb_lr)
 {
+    ovs_assert(nb_route || nb_lrp);
+
     if (route_table == NULL) {
         route_table = "";
     }
@@ -1143,9 +1145,15 @@ add_to_routes_ad(struct hmap *routes_ad, const struct in6_addr prefix,
         hmap_insert(routes_ad, &ic_route->node, hash);
     } else {
         static struct vlog_rate_limit rl = VLOG_RATE_LIMIT_INIT(5, 1);
-        VLOG_WARN_RL(&rl, "Duplicate route advertisement was suppressed! NB "
-                     "route uuid: "UUID_FMT,
-                     UUID_ARGS(&nb_route->header_.uuid));
+        const char *msg_fmt = "Duplicate %s route advertisement was "
+                              "suppressed! NB %s uuid: "UUID_FMT;
+        if (nb_route) {
+            VLOG_WARN_RL(&rl, msg_fmt, origin, "route",
+                         UUID_ARGS(&nb_route->header_.uuid));
+        } else {
+            VLOG_WARN_RL(&rl, msg_fmt, origin, "lrp",
+                         UUID_ARGS(&nb_lrp->header_.uuid));
+        }
     }
 }
 
diff --git a/include/ovn/features.h b/include/ovn/features.h
index 935db5c49..473860401 100644
--- a/include/ovn/features.h
+++ b/include/ovn/features.h
@@ -51,9 +51,11 @@ enum ovs_feature_value {
 void ovs_feature_support_destroy(void);
 bool ovs_feature_is_supported(enum ovs_feature_value feature);
 bool ovs_feature_support_run(const struct smap *ovs_capabilities,
-                             const char *br_name);
+                             const char *br_name,
+                             const char *db_target);
 bool ovs_feature_set_discovered(void);
 uint32_t ovs_feature_max_meters_get(void);
 uint32_t ovs_feature_max_select_groups_get(void);
+size_t ovs_features_max_flow_table_prefixes_get(void);
 
 #endif
diff --git a/lib/features.c b/lib/features.c
index b04437235..57d96fe3e 100644
--- a/lib/features.c
+++ b/lib/features.c
@@ -20,6 +20,9 @@
 #include "lib/util.h"
 #include "lib/dirs.h"
 #include "socket-util.h"
+#include "lib/ovsdb-cs.h"
+#include "lib/ovsdb-error.h"
+#include "lib/ovsdb-types.h"
 #include "lib/vswitch-idl.h"
 #include "odp-netlink.h"
 #include "openvswitch/vlog.h"
@@ -221,17 +224,99 @@ ovs_feature_get_openflow_cap(const char *br_name)
     return ret;
 }
 
+/* Monitoring how many prefixes flow tables can have.
+ * TODO: Remove this code once the server column type is available via IDL. */
+static struct ovsdb_cs *vswitch_cs;
+static size_t max_flow_table_prefixes;
+
+static struct json *
+vswitch_schema_prefixes_find(const struct json *schema_json,
+                             void *ctx OVS_UNUSED)
+{
+    /* We must return a monitor request object, but we do not actually want
+     * to monitor the data, so returning an empty one. */
+    struct json *monitor_request = json_object_create();
+    struct ovsdb_idl_column *prefixes_col;
+
+    prefixes_col = &ovsrec_flow_table_columns[OVSREC_FLOW_TABLE_COL_PREFIXES];
+
+    if (schema_json->type != JSON_OBJECT) {
+        VLOG_WARN_RL(&rl, "OVS database schema is not a JSON object");
+        return monitor_request;
+    }
+
+    const char *ft_name = ovsrec_table_classes[OVSREC_TABLE_FLOW_TABLE].name;
+    const char *nested_objects[] = {
+        "tables", ft_name, "columns", prefixes_col->name, "type",
+    };
+    const struct json *nested = schema_json;
+
+    for (size_t i = 0; i < ARRAY_SIZE(nested_objects); i++) {
+        nested = shash_find_data(json_object(nested), nested_objects[i]);
+        if (!nested || nested->type != JSON_OBJECT) {
+            VLOG_WARN_RL(&rl, "OVS database schema has no valid '%s'.",
+                         nested_objects[i]);
+            return monitor_request;
+        }
+    }
+
+    struct ovsdb_type server_type;
+    struct ovsdb_error *error;
+
+    error = ovsdb_type_from_json(&server_type, nested);
+    if (error) {
+        char *msg = ovsdb_error_to_string_free(error);
+
+        VLOG_WARN_RL(&rl, "Failed to parse type of %s column in %s table: %s",
+                     prefixes_col->name, ft_name, msg);
+        free(msg);
+        return monitor_request;
+    }
+
+    if (max_flow_table_prefixes != server_type.n_max) {
+        VLOG_INFO_RL(&rl, "OVS DB schema supports %d flow table prefixes, "
+                          "our IDL supports: %d",
+                     server_type.n_max, prefixes_col->type.n_max);
+        max_flow_table_prefixes = server_type.n_max;
+    }
+
+    return monitor_request;
+}
+
+static struct ovsdb_cs_ops feature_cs_ops = {
+    .compose_monitor_requests = vswitch_schema_prefixes_find,
+};
+
+static void
+vswitch_schema_prefixes_run(void)
+{
+    struct ovsdb_cs_event *event;
+    struct ovs_list events;
+
+    ovsdb_cs_run(vswitch_cs, &events);
+    LIST_FOR_EACH_POP (event, list_node, &events) {
+        /* We're not really interested in events.  We only care about the
+         * monitor request callback being invoked on re-connections or
+         * schema changes. */
+        ovsdb_cs_event_destroy(event);
+    }
+    ovsdb_cs_wait(vswitch_cs);
+}
+
 void
 ovs_feature_support_destroy(void)
 {
     rconn_destroy(swconn);
     swconn = NULL;
+    ovsdb_cs_destroy(vswitch_cs);
+    vswitch_cs = NULL;
 }
 
 /* Returns 'true' if the set of tracked OVS features has been updated. */
 bool
 ovs_feature_support_run(const struct smap *ovs_capabilities,
-                        const char *br_name)
+                        const char *br_name,
+                        const char *db_target)
 {
     static struct smap empty_caps = SMAP_INITIALIZER(&empty_caps);
     bool updated = false;
@@ -259,6 +344,16 @@ ovs_feature_support_run(const struct smap *ovs_capabilities,
                          new_state ? "supported" : "not supported");
         }
     }
+
+    if (!vswitch_cs) {
+        vswitch_cs = ovsdb_cs_create(ovsrec_idl_class.database, 3,
+                                     &feature_cs_ops, NULL);
+    }
+    ovsdb_cs_set_remote(vswitch_cs, db_target, true);
+    /* This feature doesn't affect OpenFlow rules, so not touching 'updated'
+     * even if changed. */
+    vswitch_schema_prefixes_run();
+
     return updated;
 }
 
@@ -284,3 +379,11 @@ ovs_feature_max_select_groups_get(void)
 {
     return ovs_group_features.max_groups[OFPGT11_SELECT];
 }
+
+/* Returns the number of flow table prefixes that can be configured in
+ * the OVS database. */
+size_t
+ovs_features_max_flow_table_prefixes_get(void)
+{
+    return max_flow_table_prefixes;
+}
diff --git a/lib/test-ovn-features.c b/lib/test-ovn-features.c
index aabc547e6..5ff904c12 100644
--- a/lib/test-ovn-features.c
+++ b/lib/test-ovn-features.c
@@ -26,15 +26,15 @@ test_ovn_features(struct ovs_cmdl_context *ctx OVS_UNUSED)
     struct smap features = SMAP_INITIALIZER(&features);
 
     smap_add(&features, "ct_zero_snat", "false");
-    ovs_assert(!ovs_feature_support_run(&features, NULL));
+    ovs_assert(!ovs_feature_support_run(&features, NULL, NULL));
     ovs_assert(!ovs_feature_is_supported(OVS_CT_ZERO_SNAT_SUPPORT));
 
     smap_replace(&features, "ct_zero_snat", "true");
-    ovs_assert(ovs_feature_support_run(&features, NULL));
+    ovs_assert(ovs_feature_support_run(&features, NULL, NULL));
     ovs_assert(ovs_feature_is_supported(OVS_CT_ZERO_SNAT_SUPPORT));
 
     smap_add(&features, "unknown_feature", "true");
-    ovs_assert(!ovs_feature_support_run(&features, NULL));
+    ovs_assert(!ovs_feature_support_run(&features, NULL, NULL));
 
     smap_destroy(&features);
 }
diff --git a/northd/en-lb-data.c b/northd/en-lb-data.c
index 3f5785c95..0e810eaab 100644
--- a/northd/en-lb-data.c
+++ b/northd/en-lb-data.c
@@ -181,6 +181,7 @@ lb_data_load_balancer_handler(struct engine_node *node, void *data)
             struct sset old_ips_v6 = SSET_INITIALIZER(&old_ips_v6);
             sset_swap(&lb->ips_v4, &old_ips_v4);
             sset_swap(&lb->ips_v6, &old_ips_v6);
+            enum lb_neighbor_responder_mode neigh_mode = lb->neigh_mode;
             bool routable = lb->routable;
             ovn_northd_lb_reinit(lb, tracked_lb);
             health_checks |= lb->health_checks;
@@ -215,6 +216,10 @@ lb_data_load_balancer_handler(struct engine_node *node, void *data)
                  */
                 return false;
             }
+            if (neigh_mode != lb->neigh_mode) {
+                /* If neigh_mode is updated trigger a full recompute. */
+                return false;
+            }
         }
     }
 
diff --git a/northd/northd.c b/northd/northd.c
index 3bec905cb..042c20e04 100644
--- a/northd/northd.c
+++ b/northd/northd.c
@@ -9435,78 +9435,79 @@ build_lswitch_ip_mcast_igmp_mld(struct ovn_igmp_group *igmp_group,
                                 struct ds *actions,
                                 struct ds *match)
 {
-    uint64_t dummy;
+    if (!(igmp_group->datapath && igmp_group->datapath->nbs)) {
+        return;
+    }
 
-    if (igmp_group->datapath) {
+    uint64_t dummy;
 
-        ds_clear(match);
-        ds_clear(actions);
+    ds_clear(match);
+    ds_clear(actions);
 
-        struct mcast_switch_info *mcast_sw_info =
-            &igmp_group->datapath->mcast_info.sw;
-        uint64_t table_size = mcast_sw_info->table_size;
+    struct mcast_switch_info *mcast_sw_info =
+        &igmp_group->datapath->mcast_info.sw;
+    uint64_t table_size = mcast_sw_info->table_size;
 
-        if (IN6_IS_ADDR_V4MAPPED(&igmp_group->address)) {
-            /* RFC 4541, section 2.1.2, item 2: Skip groups in the 224.0.0.X
-             * range.
-             */
-            ovs_be32 group_address =
-                in6_addr_get_mapped_ipv4(&igmp_group->address);
-            if (ip_is_local_multicast(group_address)) {
-                return;
-            }
-            if (atomic_compare_exchange_strong(
-                        &mcast_sw_info->active_v4_flows, &table_size,
-                        mcast_sw_info->table_size)) {
-                static struct vlog_rate_limit rl = VLOG_RATE_LIMIT_INIT(1, 1);
-
-                VLOG_INFO_RL(&rl, "Too many active mcast flows: %"PRIu64,
-                             mcast_sw_info->active_v4_flows);
-                return;
-            }
-            atomic_add(&mcast_sw_info->active_v4_flows, 1, &dummy);
-            ds_put_format(match, "eth.mcast && ip4 && ip4.dst == %s ",
-                          igmp_group->mcgroup.name);
-        } else {
-            /* RFC 4291, section 2.7.1: Skip groups that correspond to all
-             * hosts, all link-local routers and all site routers.
-             */
-            if (ipv6_is_all_hosts(&igmp_group->address) ||
-                ipv6_is_all_router(&igmp_group->address) ||
-                ipv6_is_all_site_router(&igmp_group->address)) {
-                return;
-            }
-            if (atomic_compare_exchange_strong(
-                        &mcast_sw_info->active_v6_flows, &table_size,
-                        mcast_sw_info->table_size)) {
-                return;
-            }
-            atomic_add(&mcast_sw_info->active_v6_flows, 1, &dummy);
-            ds_put_format(match, "eth.mcast && ip6 && ip6.dst == %s ",
-                          igmp_group->mcgroup.name);
+    if (IN6_IS_ADDR_V4MAPPED(&igmp_group->address)) {
+        /* RFC 4541, section 2.1.2, item 2: Skip groups in the 224.0.0.X
+         * range.
+         */
+        ovs_be32 group_address =
+            in6_addr_get_mapped_ipv4(&igmp_group->address);
+        if (ip_is_local_multicast(group_address)) {
+            return;
         }
+        if (atomic_compare_exchange_strong(
+                    &mcast_sw_info->active_v4_flows, &table_size,
+                    mcast_sw_info->table_size)) {
+            static struct vlog_rate_limit rl = VLOG_RATE_LIMIT_INIT(1, 1);
 
-        /* Also flood traffic to all multicast routers with relay enabled. */
-        if (mcast_sw_info->flood_relay) {
-            ds_put_cstr(actions,
-                        "clone { "
-                            "outport = \""MC_MROUTER_FLOOD "\"; "
-                            "output; "
-                        "};");
+            VLOG_INFO_RL(&rl, "Too many active mcast flows: %"PRIu64,
+                         mcast_sw_info->active_v4_flows);
+            return;
         }
-        if (mcast_sw_info->flood_static) {
-            ds_put_cstr(actions,
-                        "clone { "
-                            "outport =\""MC_STATIC"\"; "
-                            "output; "
-                        "};");
+        atomic_add(&mcast_sw_info->active_v4_flows, 1, &dummy);
+        ds_put_format(match, "eth.mcast && ip4 && ip4.dst == %s ",
+                      igmp_group->mcgroup.name);
+    } else {
+        /* RFC 4291, section 2.7.1: Skip groups that correspond to all
+         * hosts, all link-local routers and all site routers.
+         */
+        if (ipv6_is_all_hosts(&igmp_group->address) ||
+            ipv6_is_all_router(&igmp_group->address) ||
+            ipv6_is_all_site_router(&igmp_group->address)) {
+            return;
+        }
+        if (atomic_compare_exchange_strong(
+                    &mcast_sw_info->active_v6_flows, &table_size,
+                    mcast_sw_info->table_size)) {
+            return;
         }
-        ds_put_format(actions, "outport = \"%s\"; output; ",
+        atomic_add(&mcast_sw_info->active_v6_flows, 1, &dummy);
+        ds_put_format(match, "eth.mcast && ip6 && ip6.dst == %s ",
                       igmp_group->mcgroup.name);
+    }
 
-        ovn_lflow_add(lflows, igmp_group->datapath, S_SWITCH_IN_L2_LKUP,
-                      90, ds_cstr(match), ds_cstr(actions), NULL);
+    /* Also flood traffic to all multicast routers with relay enabled. */
+    if (mcast_sw_info->flood_relay) {
+        ds_put_cstr(actions,
+                    "clone { "
+                        "outport = \""MC_MROUTER_FLOOD "\"; "
+                        "output; "
+                    "};");
     }
+    if (mcast_sw_info->flood_static) {
+        ds_put_cstr(actions,
+                    "clone { "
+                        "outport =\""MC_STATIC"\"; "
+                        "output; "
+                    "};");
+    }
+    ds_put_format(actions, "outport = \"%s\"; output; ",
+                  igmp_group->mcgroup.name);
+
+    ovn_lflow_add(lflows, igmp_group->datapath, S_SWITCH_IN_L2_LKUP,
+                  90, ds_cstr(match), ds_cstr(actions), NULL);
 }
 
 /* Ingress table 25: Destination lookup, unicast handling (priority 50), */
diff --git a/tests/ovn-controller.at b/tests/ovn-controller.at
index ccffabace..4000d8261 100644
--- a/tests/ovn-controller.at
+++ b/tests/ovn-controller.at
@@ -3002,6 +3002,56 @@ OVN_CLEANUP([hv1
 /no response to inactivity probe after .* seconds, disconnecting/d])
 AT_CLEANUP
 
+AT_SETUP([ovn-controller - br-int flow table prefixes])
+AT_KEYWORDS([ovn-controller prefixes])
+ovn_start
+
+net_add n1
+sim_add hv1
+ovs-vsctl add-br br-phys
+
+dnl Stop the database server before starting ovn-controller and convert the
+dnl schema to support only 3 prefixes.
+OVS_APP_EXIT_AND_WAIT([ovsdb-server])
+AT_CHECK([sed 's/"max": 4}/"max": 3}/' $ovs_srcdir/vswitchd/vswitch.ovsschema > new_schema])
+AT_CHECK([ovsdb-tool convert hv1/conf.db new_schema])
+start_daemon ovsdb-server --remote=punix:db.sock
+
+dnl Start ovn-controller.
+ovn_attach n1 br-phys 192.168.0.20
+
+dnl Wait for exactly 3 prefixes to be configured.
+OVS_WAIT_UNTIL(
+    [grep -q 'Setting flow table prefixes: ip_src, ip_dst, ipv6_src.' \
+             hv1/ovn-controller.log])
+OVS_WAIT_FOR_OUTPUT([ovs-vsctl --columns=prefixes --bare list Flow_Table], [0], [dnl
+ip_dst ip_src ipv6_src
+])
+
+dnl Remember the row UUID.
+uuid=$(ovs-vsctl --columns=_uuid --bare list Flow_Table)
+
+dnl Stop the database server again and convert the schema back to support
+dnl up to 4 prefixes.
+OVS_APP_EXIT_AND_WAIT([ovsdb-server])
+AT_CHECK([cat $ovs_srcdir/vswitchd/vswitch.ovsschema > new_schema])
+AT_CHECK([ovsdb-tool convert hv1/conf.db new_schema])
+start_daemon ovsdb-server --remote=punix:db.sock
+
+dnl Wait for exactly 4 prefixes to be configured.
+OVS_WAIT_FOR_OUTPUT(
+    [grep -q 'Setting flow table prefixes: ip_src, ip_dst, ipv6_src, ipv6_dst.' \
+             hv1/ovn-controller.log])
+OVS_WAIT_FOR_OUTPUT([ovs-vsctl --columns=prefixes --bare list Flow_Table], [0], [dnl
+ip_dst ip_src ipv6_dst ipv6_src
+])
+
+dnl Check that the record was updated and not replaced.
+AT_CHECK([test "$(ovs-vsctl --columns=_uuid --bare list Flow_Table)" = "${uuid}"])
+
+OVN_CLEANUP([hv1])
+AT_CLEANUP
+
 AT_SETUP([ovn-controller - Start controller with empty db])
 AT_KEYWORDS([ovn])
 
diff --git a/tests/ovn-ic.at b/tests/ovn-ic.at
index 70efcdf2c..bf664b55e 100644
--- a/tests/ovn-ic.at
+++ b/tests/ovn-ic.at
@@ -282,6 +282,67 @@ OVN_CLEANUP_IC
 AT_CLEANUP
 ])
 
+OVN_FOR_EACH_NORTHD([
+AT_SETUP([ovn-ic -- local duplicate connected route])
+
+ovn_init_ic_db
+net_add n1
+
+# 1 GW per AZ
+for i in 1 2; do
+    az=az$i
+    ovn_start $az
+    sim_add gw-$az
+    as gw-$az
+    check ovs-vsctl add-br br-phys
+    ovn_az_attach $az n1 br-phys 192.168.1.$i
+    check ovs-vsctl set open . external-ids:ovn-is-interconn=true
+    check ovn-nbctl set nb-global . \
+        options:ic-route-adv=true \
+        options:ic-route-adv-default=true \
+        options:ic-route-learn=true \
+        options:ic-route-learn-default=true
+done
+
+ovn_as az1
+
+# create transit switch and connect to LR
+check ovn-ic-nbctl --wait=sb ts-add ts1
+for i in 1 2; do
+    ovn_as az$i
+
+    check \
+       grep -vq "connected route advertisement was suppressed! NB lrp" \
+       az$i/ic/ovn-ic.log
+
+    check ovn-nbctl lr-add lr1
+    check ovn-nbctl lrp-add lr1 lrp$i 00:00:00:00:0$i:01 10.0.$i.1/24
+    check ovn-nbctl lrp-add lr1 lrp-dup1 00:00:00:00:0$i:42 10.0.42.1/24
+    check ovn-nbctl lrp-add lr1 lrp-dup2 00:00:00:00:0$i:51 10.0.42.1/24
+    check ovn-nbctl lrp-set-gateway-chassis lrp$i gw-az$i
+
+    check ovn-nbctl --wait=sb lsp-add ts1 lsp$i -- \
+        lsp-set-addresses lsp$i router -- \
+        lsp-set-type lsp$i router -- \
+        lsp-set-options lsp$i router-port=lrp$i
+    check ovn-ic-nbctl --wait=sb sync
+
+    check \
+       grep -q "connected route advertisement was suppressed! NB lrp" \
+       az$i/ic/ovn-ic.log
+done
+
+for i in 1 2; do
+    az=az$i
+    ovn_as $az
+    OVN_CLEANUP_SBOX(gw-$az)
+    OVN_CLEANUP_AZ([$az])
+done
+
+OVN_CLEANUP_IC
+AT_CLEANUP
+])
+
 OVN_FOR_EACH_NORTHD([
 AT_SETUP([ovn-ic -- gateway sync])
 
diff --git a/tests/ovn-macros.at b/tests/ovn-macros.at
index a08252d50..fb3d621d0 100644
--- a/tests/ovn-macros.at
+++ b/tests/ovn-macros.at
@@ -937,8 +937,7 @@ send_garp() {
                       ARP(op=$op, hwsrc='${eth_src}', hwdst='${eth_src}', \
                       psrc='${spa}', pdst='${tpa}')")
     echo "Sending GARP($op) $eth_src $eth_dst $spa $tpa packet=$packet on $hv"
-    as $hv
-    ovs-appctl netdev-dummy/receive $inport $packet
+    check as $hv ovs-appctl netdev-dummy/receive $inport $packet
 }
 
 # Wrapper on top of ovn-trace, stripping some things and storing the trace
diff --git a/tests/ovn-northd.at b/tests/ovn-northd.at
index 83a3d1549..c5d097b4c 100644
--- a/tests/ovn-northd.at
+++ b/tests/ovn-northd.at
@@ -10993,8 +10993,35 @@ check_engine_stats lr_stateful norecompute compute
 check_engine_stats lflow norecompute compute
 check_engine_stats sync_to_sb_lb norecompute compute
 CHECK_NO_CHANGE_AFTER_RECOMPUTE
+
+check as northd ovn-appctl -t ovn-northd inc-engine/clear-stats
+check ovn-nbctl --wait=sb set load_balancer lb1 options:neighbor_responder=all
+check_engine_stats lb_data recompute nocompute
+check_engine_stats northd recompute nocompute
+check_engine_stats lr_stateful recompute nocompute
+check_engine_stats lflow recompute nocompute
+check_engine_stats sync_to_sb_lb recompute nocompute
+CHECK_NO_CHANGE_AFTER_RECOMPUTE
+
+check as northd ovn-appctl -t ovn-northd inc-engine/clear-stats
+check ovn-nbctl --wait=sb set load_balancer lb1 options:neighbor_responder=reachable
+check_engine_stats lb_data recompute nocompute
+check_engine_stats northd recompute nocompute
+check_engine_stats lr_stateful recompute nocompute
+check_engine_stats lflow recompute nocompute
+check_engine_stats sync_to_sb_lb recompute nocompute
+CHECK_NO_CHANGE_AFTER_RECOMPUTE
+
 check as northd ovn-appctl -t ovn-northd inc-engine/clear-stats
+check ovn-nbctl --wait=sb set load_balancer lb1 options:neighbor_responder=none
+check_engine_stats lb_data recompute nocompute
+check_engine_stats northd recompute nocompute
+check_engine_stats lr_stateful recompute nocompute
+check_engine_stats lflow recompute nocompute
+check_engine_stats sync_to_sb_lb recompute nocompute
+CHECK_NO_CHANGE_AFTER_RECOMPUTE
 
+check as northd ovn-appctl -t ovn-northd inc-engine/clear-stats
 check ovn-nbctl --wait=sb -- lb-add lb2 20.0.0.10:80 20.0.0.20:80 -- lb-add lb3 30.0.0.10:80 30.0.0.20:80
 check_engine_stats lb_data norecompute compute
 check_engine_stats northd norecompute compute
diff --git a/tests/ovn.at b/tests/ovn.at
index 28c96b08b..a219d1e94 100644
--- a/tests/ovn.at
+++ b/tests/ovn.at
@@ -13476,7 +13476,7 @@ AT_CHECK([
 
 spa=$(ip_to_hex 10 0 0 1)
 tpa=$(ip_to_hex 10 0 0 100)
-send_garp hv1 1 1 "00:00:00:00:00:01" "ff:ff:ff:ff:ff:ff" $spa $tpa
+send_garp hv1 vif1 1 "00:00:00:00:00:01" "ff:ff:ff:ff:ff:ff" $spa $tpa
 
 dnl traffic from localport should not be sent to localnet
 AT_CHECK([tcpdump -r hv1/br-phys_n1-tx.pcap arp[[24:4]]=0x0a000064 | wc -l],[0],[dnl
@@ -34583,6 +34583,7 @@ ovs-vsctl -- add-port br-phys ext1 -- \
     options:rxq_pcap=hv1/ext1-rx.pcap \
     ofport-request=2
 ovs-vsctl set open . external_ids:ovn-bridge-mappings=physnet1:br-phys
+ovn-appctl -t ovn-controller vlog/set mac_cache:file:dbg pinctrl:file:dbg
 
 sim_add hv2
 as hv2
@@ -34600,6 +34601,7 @@ ovs-vsctl -- add-port br-phys ext2 -- \
     options:rxq_pcap=hv2/ext2-rx.pcap \
     ofport-request=2
 ovs-vsctl set open . external_ids:ovn-bridge-mappings=physnet1:br-phys
+ovn-appctl -t ovn-controller vlog/set mac_cache:file:dbg pinctrl:file:dbg
 
 OVN_POPULATE_ARP
 wait_for_ports_up
@@ -34643,6 +34645,7 @@ AT_CHECK_UNQUOTED([as hv1 ovs-ofctl dump-flows br-int table=79 --no-stats | stri
 ])
 
 timestamp=$(fetch_column mac_binding timestamp ip="192.168.10.20")
+binding_uuid=$(fetch_column mac_binding _uuid ip="192.168.10.20")
 
 send_udp hv1 ext1 10
 send_udp hv2 ext2 20
@@ -34665,7 +34668,10 @@ OVS_CTL_TIMEOUT=10
 OVS_WAIT_UNTIL([
     test "$timestamp" != "$(fetch_column mac_binding timestamp ip='192.168.10.20')"
 ])
-check $(test "$(fetch_column mac_binding timestamp ip='192.168.10.20')" != "")
+# Should still have 2 entries, i.e. none should expire yet.
+check_row_count mac_binding 2 ip='192.168.10.20'
+# Must be the same updated entries, not re-created.
+check test "$binding_uuid" = "$(fetch_column mac_binding _uuid ip='192.168.10.20')"
 
 # Check if the records are removed after some inactivity for gw-1. Only 1 entry should be present for gw-2.
 OVS_WAIT_UNTIL([
@@ -34690,14 +34696,12 @@ AT_CHECK([ovn-sbctl get datapath $uuid external_ids:mac_binding_age_threshold],
 ])
 
 # Send GARP to populate MAC binding table records
-send_garp hv1 ext1 10 # belong to 192.168.10.0/24
-send_garp hv2 ext2 20 # belong to 192.168.10.20/32
-send_garp hv2 ext2 65 # belong to 192.168.10.64/26
-
-ip_addresses=("192.168.10.10" "192.168.10.20" "192.168.10.65")
-logical_ports=("gw-1-public" "gw-2-public")
-for ip in "${ip_addresses[@]}"; do
-    for port in "${logical_ports[@]}"; do
+send_garp hv1 ext1 2 "00:00:00:00:10:10" "ff:ff:ff:ff:ff:ff" "192.168.10.10" "192.168.10.10" # belong to 192.168.10.0/24
+send_garp hv2 ext2 2 "00:00:00:00:10:20" "ff:ff:ff:ff:ff:ff" "192.168.10.20" "192.168.10.20" # belong to 192.168.10.20/32
+send_garp hv2 ext2 2 "00:00:00:00:10:65" "ff:ff:ff:ff:ff:ff" "192.168.10.65" "192.168.10.65" # belong to 192.168.10.64/26
+
+for ip in 192.168.10.10 192.168.10.20 192.168.10.65; do
+    for port in gw-1-public gw-2-public; do
         wait_row_count mac_binding 1 ip=$ip logical_port=$port
     done
 done
@@ -34707,9 +34711,8 @@ OVS_WAIT_UNTIL([
 ])
 
 # The other two should remain because the corresponding prefixes have threshold 0
-ip_addresses=("192.168.10.20" "192.168.10.65")
-for ip in "${ip_addresses[@]}"; do
-    for port in "${logical_ports[@]}"; do
+for ip in 192.168.10.20 192.168.10.65; do
+    for port in gw-1-public gw-2-public; do
         check_row_count mac_binding 1 ip=$ip logical_port=$port
     done
 done
@@ -34725,11 +34728,11 @@ AT_CHECK([ovn-sbctl get datapath $uuid external_ids:mac_binding_age_threshold],
 ])
 
 # Send GARP to populate MAC binding table records
-send_garp hv1 ext1 10 # belong to default
-send_garp hv2 ext2 65 # belong to 192.168.10.64/26
+send_garp hv1 ext1 2 "00:00:00:00:10:10" "ff:ff:ff:ff:ff:ff" "192.168.10.10" "192.168.10.10" # belong to default
+send_garp hv2 ext2 2 "00:00:00:00:10:65" "ff:ff:ff:ff:ff:ff" "192.168.10.65" "192.168.10.65" # belong to 192.168.10.64/26
 
-for ip in "${ip_addresses[@]}"; do
-    for port in "${logical_ports[@]}"; do
+for ip in 192.168.10.10 192.168.10.65; do
+    for port in gw-1-public gw-2-public; do
         wait_row_count mac_binding 1 ip=$ip logical_port=$port
     done
 done
@@ -34737,7 +34740,7 @@ done
 OVS_WAIT_UNTIL([
     test "1" = "$(ovn-sbctl list mac_binding | grep -c '192.168.10.10')"
 ])
-for port in "${logical_ports[@]}"; do
+for port in gw-1-public gw-2-public; do
     check_row_count mac_binding 1 ip="192.168.10.65" logical_port=$port
 done
 check ovn-sbctl --all destroy mac_binding
@@ -34749,7 +34752,7 @@ uuid=$(fetch_column datapath _uuid external_ids:name=gw-1)
 AT_CHECK([ovn-sbctl get datapath $uuid external_ids:mac_binding_age_threshold], [1], [ignore], [ignore])
 
 # Send GARP to populate MAC binding table records
-send_garp hv1 ext1 10
+send_garp hv1 ext1 2 "00:00:00:00:10:10" "ff:ff:ff:ff:ff:ff" "192.168.10.10" "192.168.10.10"
 
 wait_row_count mac_binding 1 ip="192.168.10.10" logical_port="gw-1-public"
 wait_row_count mac_binding 1 ip="192.168.10.10" logical_port="gw-2-public"
@@ -34773,18 +34776,15 @@ AT_CHECK([ovn-sbctl get datapath $uuid external_ids:mac_binding_age_threshold],
 ])
 
 # Send GARP to populate MAC binding table records
-send_garp hv1 ext1 10 # belong to 192.168.10.0/24
-send_garp hv2 ext2 20 # belong to 192.168.10.20/32
+send_garp hv1 ext1 2 "00:00:00:00:10:10" "ff:ff:ff:ff:ff:ff" "192.168.10.10" "192.168.10.10" # belong to 192.168.10.0/24
+send_garp hv2 ext2 2 "00:00:00:00:10:20" "ff:ff:ff:ff:ff:ff" "192.168.10.20" "192.168.10.20" # belong to 192.168.10.20/32
 
-ip_addresses=("192.168.10.10" "192.168.10.20")
-logical_ports=("gw-1-public" "gw-2-public")
-for ip in "${ip_addresses[@]}"; do
-    for port in "${logical_ports[@]}"; do
+for ip in 192.168.10.10 192.168.10.20; do
+    for port in gw-1-public gw-2-public; do
         wait_row_count mac_binding 1 ip=$ip logical_port=$port
     done
 done
 
-
 OVS_WAIT_UNTIL([
     test "0" = "$(ovn-sbctl list mac_binding | grep -c '192.168.10.10')"
 ])
@@ -34806,6 +34806,7 @@ sim_add hv1
 as hv1
 check ovs-vsctl add-br br-phys
 ovn_attach n1 br-phys 192.168.0.1
+ovn-appctl -t ovn-controller vlog/set mac_cache:file:dbg pinctrl:file:dbg
 
 check ovn-nbctl                                                     \
     -- ls-add ls1                                                   \
@@ -34847,10 +34848,10 @@ wait_row_count mac_binding 1 ip="192.168.10.10" logical_port="lr-ls1"
 
 dnl Check local mac binding cache.
 lr_key=$(fetch_column datapath_binding tunnel_key external_ids:name=lr)
-mb_uuid=$(fetch_column mac_binding _uuid logical_port=lr-ls1)
+cookie=$(fetch_column mac_binding _uuid logical_port=lr-ls1 | cut -d'-' -f1)
 OVS_WAIT_FOR_OUTPUT_UNQUOTED([as hv1 ovn-appctl -t ovn-controller debug/dump-mac-bindings], [0], [dnl
 Local MAC bindings:
-SB UUID: $mb_uuid, datapath-key: $lr_key, port-key: 0, ip: 192.168.10.10, mac: 00:00:00:00:10:10
+cookie: 0x$cookie, datapath-key: $lr_key, port-key: 0, ip: 192.168.10.10, mac: 00:00:00:00:10:10
 ])
 
 dnl Remove router port.  This deletes the mac binding in the same transaction.
@@ -34873,10 +34874,10 @@ send_garp hv1 vif2 2 00:00:00:00:20:10 ff:ff:ff:ff:ff:ff 192.168.20.10 192.168.2
 wait_row_count mac_binding 1 ip="192.168.20.10" logical_port="lr-ls2"
 
 dnl Check local mac binding cache.
-mb_uuid=$(fetch_column mac_binding _uuid logical_port=lr-ls2)
+cookie=$(fetch_column mac_binding _uuid logical_port=lr-ls2 | cut -d'-' -f1)
 OVS_WAIT_FOR_OUTPUT_UNQUOTED([as hv1 ovn-appctl -t ovn-controller debug/dump-mac-bindings], [0], [dnl
 Local MAC bindings:
-SB UUID: $mb_uuid, datapath-key: $lr_key, port-key: 0, ip: 192.168.20.10, mac: 00:00:00:00:20:10
+cookie: 0x$cookie, datapath-key: $lr_key, port-key: 0, ip: 192.168.20.10, mac: 00:00:00:00:20:10
 ])
 
 dnl Change LRP tunnel key.
@@ -34894,6 +34895,89 @@ OVN_CLEANUP([hv1])
 AT_CLEANUP
 ])
 
+OVN_FOR_EACH_NORTHD([
+AT_SETUP([MAC binding aging - persistence of the active entry])
+AT_SKIP_IF([test $HAVE_SCAPY = no])
+ovn_start
+
+net_add n1
+sim_add hv1
+as hv1
+check ovs-vsctl add-br br-phys
+ovn_attach n1 br-phys 192.168.0.1
+ovn-appctl -t ovn-controller vlog/set mac_cache:file:dbg pinctrl:file:dbg
+
+check ovn-nbctl                                                     \
+    -- ls-add ls1                                                   \
+    -- ls-add ls2                                                   \
+    -- lr-add lr                                                    \
+    -- set logical_router lr options:mac_binding_age_threshold=4    \
+    -- lrp-add lr lr-ls1 00:00:00:00:10:00 192.168.10.1/24          \
+    -- lrp-add lr lr-ls2 00:00:00:00:20:00 192.168.20.1/24          \
+    -- lsp-add ls1 ls1-lr                                           \
+    -- lsp-set-type ls1-lr router                                   \
+    -- lsp-set-addresses ls1-lr router                              \
+    -- lsp-set-options ls1-lr router-port=lr-ls1                    \
+    -- lsp-add ls1 vif1                                             \
+    -- lsp-set-addresses vif1 "00:00:00:00:10:10 192.168.10.10"     \
+    -- lsp-add ls2 ls2-lr                                           \
+    -- lsp-set-type ls2-lr router                                   \
+    -- lsp-set-addresses ls2-lr router                              \
+    -- lsp-set-options ls2-lr router-port=lr-ls2                    \
+    -- lsp-add ls2 vif2                                             \
+    -- lsp-set-addresses vif2 "00:00:00:00:20:10 192.168.20.10"
+
+check ovs-vsctl                                      \
+    -- add-port br-int vif1                          \
+    -- set interface vif1 external-ids:iface-id=vif1 \
+    -- add-port br-int vif2                          \
+    -- set interface vif2 external-ids:iface-id=vif2
+
+OVN_POPULATE_ARP
+wait_for_ports_up
+check ovn-nbctl --wait=hv sync
+
+dnl Wait for pinctrl thread to be connected.
+OVS_WAIT_UNTIL([grep pinctrl hv1/ovn-controller.log | grep -q connected])
+
+dnl Populate MAC Binding entries.  Wait between two, so they have an initial
+dnl difference in timestamps around half the age threshold.
+send_garp hv1 vif2 2 00:00:00:00:20:10 ff:ff:ff:ff:ff:ff 192.168.20.10 192.168.20.10
+wait_row_count mac_binding 1 ip="192.168.20.10" logical_port="lr-ls2"
+sleep 2
+send_garp hv1 vif1 2 00:00:00:00:10:10 ff:ff:ff:ff:ff:ff 192.168.10.10 192.168.10.10
+wait_row_count mac_binding 1 ip="192.168.10.10" logical_port="lr-ls1"
+
+send_udp() {
+    local hv=$1 dev=$2 byte=$3
+
+    packet=$(fmt_pkt "Ether(dst='00:00:00:00:10:00', src='00:00:00:00:10:${byte}')/ \
+                      IP(dst='192.168.20.${byte}', src='192.168.10.${byte}')/UDP()")
+    check as $hv ovs-appctl netdev-dummy/receive $dev $packet
+}
+
+uuid=$(fetch_column mac_binding _uuid ip="192.168.10.10" logical_port="lr-ls1")
+for i in $(seq 12); do
+    # Keep one entry alive by sending traffic that uses it.
+    send_udp hv1 vif1 10
+    sleep 1
+    # The entry must not expire.
+    check_row_count mac_binding 1 ip="192.168.10.10" logical_port="lr-ls1"
+    as hv1 ovs-ofctl dump-flows br-int table=OFTABLE_MAC_CACHE_USE
+done
+# Check that it's the same entry.
+uuid2=$(fetch_column mac_binding _uuid ip="192.168.10.10" logical_port="lr-ls1")
+check test "$uuid" = "$uuid2"
+
+# The other entry must have expired by now.
+check_row_count mac_binding 0 ip="192.168.20.10" logical_port="lr-ls2"
+# Wait for the alive one to expire as well.
+wait_row_count mac_binding 0 ip="192.168.10.10" logical_port="lr-ls1"
+
+OVN_CLEANUP([hv1])
+AT_CLEANUP
+])
+
 OVN_FOR_EACH_NORTHD([
 AT_SETUP([router port type update and then remove])
 ovn_start
@@ -36833,6 +36917,7 @@ ovs-vsctl -- add-port br-phys ext0 -- \
     options:rxq_pcap=hv1/ext0-rx.pcap \
     ofport-request=2
 ovs-vsctl set open . external_ids:ovn-bridge-mappings=physnet1:br-phys
+ovn-appctl -t ovn-controller vlog/set mac_cache:file:dbg pinctrl:file:dbg
 
 OVN_POPULATE_ARP
 wait_for_ports_up
@@ -36878,6 +36963,88 @@ OVN_CLEANUP([hv1])
 AT_CLEANUP
 ])
 
+OVN_FOR_EACH_NORTHD([
+AT_SETUP([FDB aging - persistence of the active entry])
+AT_SKIP_IF([test $HAVE_SCAPY = no])
+ovn_start
+
+net_add n1
+
+check ovn-nbctl ls-add ls0
+
+check ovn-nbctl lsp-add ls0 ln_port -- \
+      lsp-set-addresses ln_port unknown -- \
+      lsp-set-type ln_port localnet -- \
+      lsp-set-options ln_port network_name=physnet1 -- \
+      set logical_switch_port ln_port options:localnet_learn_fdb=true -- \
+      set logical_switch ls0 other_config:fdb_age_threshold=4
+
+check ovn-nbctl lsp-add ls0 vif1 -- \
+      lsp-set-addresses vif1 "00:00:00:00:10:10 192.168.10.10"
+
+sim_add hv1
+as hv1
+ovs-vsctl add-br br-underlay
+ovn_attach n1 br-underlay 192.168.0.1
+ovs-vsctl add-br br-phys
+ovs-vsctl -- add-port br-int vif1 -- \
+    set interface vif1 external-ids:iface-id=vif1 \
+    options:tx_pcap=hv1/vif1-tx.pcap \
+    options:rxq_pcap=hv1/vif1-rx.pcap \
+    ofport-request=1
+ovs-vsctl -- add-port br-phys ext0 -- \
+    set interface ext0 \
+    options:tx_pcap=hv1/ext0-tx.pcap \
+    options:rxq_pcap=hv1/ext0-rx.pcap \
+    ofport-request=2
+ovs-vsctl set open . external_ids:ovn-bridge-mappings=physnet1:br-phys
+ovn-appctl -t ovn-controller vlog/set mac_cache:file:dbg pinctrl:file:dbg
+
+OVN_POPULATE_ARP
+wait_for_ports_up
+check ovn-nbctl --wait=hv sync
+
+send_packet() {
+    packet=$(fmt_pkt "
+            Ether(dst='00:00:00:00:10:10', src='00:00:00:00:10:${1}') /
+            IP(src='192.168.10.${1}', dst='192.168.10.10') /
+            UDP(sport=1234, dport=1235)
+           ")
+    check ovs-appctl netdev-dummy/receive ext0 $packet
+}
+
+# Send packets to create FDB entries.  Wait between two, so they have an
+# initial difference in timestamps around half the age threshold.
+send_packet 20
+wait_row_count fdb 1 mac='"00:00:00:00:10:20"'
+sleep 2
+send_packet 30
+wait_row_count fdb 1 mac='"00:00:00:00:10:30"'
+
+timestamp=$(fetch_column fdb timestamp mac='"00:00:00:00:10:20"')
+
+uuid=$(fetch_column fdb _uuid mac='"00:00:00:00:10:30"')
+for i in $(seq 12); do
+    # Keep one entry alive by sending traffic that uses it.
+    send_packet 30
+    sleep 1
+    # The entry must not expire.
+    check_row_count fdb 1 mac='"00:00:00:00:10:30"'
+    as hv1 ovs-ofctl dump-flows br-int table=OFTABLE_LOOKUP_FDB
+done
+# Check that it's the same entry.
+uuid2=$(fetch_column fdb _uuid mac='"00:00:00:00:10:30"')
+check test "$uuid" = "$uuid2"
+
+# The other entry must have expired by now.
+wait_row_count fdb 0 mac='"00:00:00:00:10:20"'
+# Wait for the alive one to expire as well.
+wait_row_count fdb 0 mac='"00:00:00:00:10:30"'
+
+OVN_CLEANUP([hv1])
+AT_CLEANUP
+])
+
 OVN_FOR_EACH_NORTHD([
 AT_SETUP([DNAT_SNAT and LB traffic])
 AT_KEYWORDS([dnat-snat-lb])
diff --git a/tests/system-ovn.at b/tests/system-ovn.at
index 52101f96d..349da2fcb 100644
--- a/tests/system-ovn.at
+++ b/tests/system-ovn.at
@@ -12778,3 +12778,72 @@ OVS_TRAFFIC_VSWITCHD_STOP(["/failed to query port patch-.*/d
 /connection dropped.*/d"])
 AT_CLEANUP
 ])
+
+OVN_FOR_EACH_NORTHD([
+AT_SETUP([neighbor_responder for all])
+AT_KEYWORDS([neighbor_responder])
+AT_SKIP_IF([test "$HAVE_ARPING" = no])
+
+CHECK_CONNTRACK()
+CHECK_CONNTRACK_NAT()
+
+ovn_start
+OVS_TRAFFIC_VSWITCHD_START()
+ADD_BR([br-int])
+
+# Set external-ids in br-int needed for ovn-controller
+ovs-vsctl \
+        -- set Open_vSwitch . external-ids:system-id=hv1 \
+        -- set Open_vSwitch . external-ids:ovn-remote=unix:$ovs_base/ovn-sb/ovn-sb.sock \
+        -- set Open_vSwitch . external-ids:ovn-encap-type=geneve \
+        -- set Open_vSwitch . external-ids:ovn-encap-ip=169.0.0.1 \
+        -- set bridge br-int fail-mode=secure other-config:disable-in-band=true
+
+# Start ovn-controller
+start_daemon ovn-controller
+
+check_uuid ovn-nbctl create Logical_Router name=lr options:chassis=hv1
+check ovn-nbctl ls-add ls
+
+# Connect alice to R2
+check ovn-nbctl lrp-add lr lr-ls 00:00:02:01:02:03 192.168.1.254/24
+check ovn-nbctl lsp-add ls ls-lr -- set Logical_Switch_Port ls-lr \
+    type=router options:router-port=lr-ls addresses=\"00:00:02:01:02:03\"
+
+# Logical port 'foo' in switch 'foo'.
+ADD_NAMESPACES(foo)
+ADD_VETH(foo, foo, br-int, "192.168.1.1/24", "f0:00:00:01:02:03", \
+         "192.168.1.254")
+check ovn-nbctl lsp-add ls foo \
+-- lsp-set-addresses foo "f0:00:00:01:02:03 192.168.1.1"
+
+check ovn-nbctl lb-add lb0 192.168.1.100 192.168.1.1
+check ovn-nbctl lr-lb-add lr lb0
+check ovn-nbctl lb-add lb1 172.16.1.100 192.168.1.1
+check ovn-nbctl lr-lb-add lr lb1
+
+NS_CHECK_EXEC([foo], [ip route add 172.16.1.100 dev foo])
+
+check ovn-nbctl set load_balancer lb0 options:neighbor_responder=all
+check ovn-nbctl set load_balancer lb1 options:neighbor_responder=all
+check ovn-nbctl --wait=hv sync
+
+NS_CHECK_EXEC([foo], [arping -q -c 3 192.168.1.100])
+NS_CHECK_EXEC([foo], [arping -q -c 3 172.16.1.100])
+
+OVS_APP_EXIT_AND_WAIT([ovn-controller])
+
+as ovn-sb
+OVS_APP_EXIT_AND_WAIT([ovsdb-server])
+
+as ovn-nb
+OVS_APP_EXIT_AND_WAIT([ovsdb-server])
+
+as northd
+OVS_APP_EXIT_AND_WAIT([ovn-northd])
+
+as
+OVS_TRAFFIC_VSWITCHD_STOP(["/failed to query port patch-.*/d
+/connection dropped.*/d"])
+AT_CLEANUP
+])
