diff --git a/.github/workflows/test.yml b/.github/workflows/test.yml
index 0cb981775..456ab5c69 100644
--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -225,6 +225,14 @@ jobs:
         sudo chmod +x /usr/bin/crun
         echo "New crun version: "$(crun --version)
 
+    - name: Reduce ASLR entropy
+      if:   matrix.cfg.sanitizers != ''
+      # Asan in llvm 14 provided in ubuntu-22.04 is incompatible with
+      # high-entropy ASLR configured in much newer kernels that GitHub
+      # runners are using leading to random crashes:
+      #   https://github.com/actions/runner-images/issues/9491
+      run: sudo sysctl -w vm.mmap_rnd_bits=28
+
     - name: load image
       run: |
         sudo podman load -i /tmp/image.tar
diff --git a/NEWS b/NEWS
index e91cbe43f..ca3562425 100644
--- a/NEWS
+++ b/NEWS
@@ -1,3 +1,6 @@
+OVN v24.03.2 - xx xxx xxxx
+--------------------------
+
 OVN v24.03.1 - 12 Mar 2024
 --------------------------
   - Bug fixes
diff --git a/configure.ac b/configure.ac
index 5f15422f2..962422bd2 100644
--- a/configure.ac
+++ b/configure.ac
@@ -13,7 +13,7 @@
 # limitations under the License.
 
 AC_PREREQ(2.63)
-AC_INIT(ovn, 24.03.1, bugs@openvswitch.org)
+AC_INIT(ovn, 24.03.2, bugs@openvswitch.org)
 AC_CONFIG_MACRO_DIR([m4])
 AC_CONFIG_AUX_DIR([build-aux])
 AC_CONFIG_HEADERS([config.h])
diff --git a/debian/changelog b/debian/changelog
index 2f20941be..acf278a15 100644
--- a/debian/changelog
+++ b/debian/changelog
@@ -1,3 +1,9 @@
+OVN (24.03.2-1) unstable; urgency=low
+   [ OVN team ]
+   * New upstream version
+
+ -- OVN team <dev@openvswitch.org>  Tue, 12 Mar 2024 08:34:38 -0400
+
 OVN (24.03.1-1) unstable; urgency=low
    [ OVN team ]
    * New upstream version
